src/components/PingChart.vue

```diff
@@ -226,42 +404,125 @@ export default {
+    beforeUnmount() {
+        if (this.chartDataFetchInterval) {
+            clearInterval(this.chartDataFetchInterval);
         }
+    },
+    methods: {
+        // Get color of bar chart for this datapoint
+        getBarColorForDatapoint(datapoint) {
+            if (datapoint.maintenance != null) {
+                // Target is in maintenance
+                return "rgba(23,71,245,0.41)";
+            } else if (datapoint.down === 0) {
+                // Target is up
+                return "#FFFFFFFF";
+            } else if (datapoint.up === 0) {
+                // Target is down
+                return "#DC354568";
+            } else {
+                // Show yellow for mixed status
+                return "rgba(245,182,23,0.41)";
+            }
+        },
+        // push datapoint to chartData
+        pushDatapoint(datapoint, avgPingData, minPingData, maxPingData, downData, colorData) {
+            const x = this.$root.unixToDateTime(datapoint.timestamp);
+
+            // Show ping values if it was up in this period
+            avgPingData.push({
+                x,
+                y: datapoint.up > 0 && datapoint.avgPing > 0 ? datapoint.avgPing : null,
+            });
+            minPingData.push({
+                x,
+                y: datapoint.up > 0 && datapoint.avgPing > 0 ? datapoint.minPing : null,
+            });
+            maxPingData.push({
+                x,
+                y: datapoint.up > 0 && datapoint.avgPing > 0 ? datapoint.maxPing : null,
+            });
+
+            let bar = {
+                x,
+                y: datapoint.down,
+            };
+
+            if (datapoint.maintenance) {
+                bar.y += datapoint.maintenance;
+            }
+
+            downData.push(bar);
+
+            colorData.push(this.getBarColorForDatapoint(datapoint));
+        },
+        // get the average of a set of datapoints
+        getAverage(datapoints) {
+            const totalUp = datapoints.reduce((total, current) => total + current.up, 0);
+            const totalDown = datapoints.reduce((total, current) => total + current.down, 0);
+            const totalMaintenance = datapoints.reduce((total, current) => total + (current.maintenance || 0), 0);
+            const totalPing = datapoints.reduce((total, current) => total + current.avgPing * current.up, 0);
+            const minPing = datapoints.reduce((min, current) => Math.min(min, current.minPing), Infinity);
+            const maxPing = datapoints.reduce((max, current) => Math.max(max, current.maxPing), 0);
+
+            // Find the middle timestamp to use
+            let midpoint = 0;
+            if (datapoints.length > 1) {
+                midpoint = Math.floor(datapoints.length / 2);
+            }
+
+            return {
+                timestamp: datapoints[midpoint].timestamp,
+                up: totalUp,
+                down: totalDown,
+                maintenance: totalMaintenance > 0 ? totalMaintenance : undefined,
```

Comment:
I'm not sure what you're asking.

Reply:
Why `maintenance` is `undefined` instead of `0`.
I know that this is to match the backend, but I don't quite get why the functionality is like this in the backend.

Reply:
maintenance is stored as json data in an `extras` column. It makes no sense to keep storing `maintenance: 0` as that would just waste space, and indeed it was implemented to not do that. That said, aligning the data format here is more just code style and should have no impact on the functionality.